### PR TITLE
Stackdriver Instrumentation BA-4788

### PR DESCRIPTION
**Pull request by salonishah11**
This adds the option of sending metrics to Google's Stackdriver API. 

Closes [JIRA BA-4788](https://broadworkbench.atlassian.net/browse/BA-4788).

### CODE REVIEW

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    case ShutdownCommand =>
+      // flush out metrics (if any) before shut down
+      sendMetricData()
+      metricServiceClient.close()
+      context stop self
+  }
+
+
+  private def labelFromConfig(op: StackdriverConfig => Option[String], key: String): List[(String, String)] = {
+    op(stackdriverConfig).fold(List.empty[(String, String)])(v => List((key.replace("-", "_"), v)))
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: It seems weird to me that you're returning a list here, since if you returned an `Option` this could just be:
```scala
def labelFromConfig(op: StackdriverConfig => Option[String], key: String): Option[(String, String)] = 
  op(stackdriverConfig).map(v => List((key.replace("-", "_"), v)))
```

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+
+  private def generateMetricLabels(): Map[String, String] = {
+    val labelsList = labelFromConfig(_.cromwellInstanceIdentifier, CromwellInstanceIdentifier) ++
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: if `labelFromConfig ` returned an `Option` (from my comment above) you could go straight to a map by using `toMap` on the first element, eg:
```scala
labelFromConfig(_.cromwellInstanceIdentifier, CromwellInstanceIdentifier).toMap ++
  labelFromConfig(_.cromwellInstanceRole, CromwellInstanceRole) ++
  labelFromConfig(_.cromwellPerfTestCase, CromwellPerfTest)
```

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        case StackdriverCumulative => dataPointListSum
+      }
+      writeMetrics(key, dataPoint)
+    }
+
+    metricsMap = Map.empty[StackdriverMetric, List[Double]]
+  }
+
+
+  private def timeInterval(metricKind: StackdriverMetricKind): TimeInterval = {
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: TOL: For these three helper methods (`timeInterval`, `createTimeSeries` and `createMetricBuilder`) which are only ever used in `writeMetrics`, my preference is to move them inside that method so that it's obvious that they only exist to help that method do its work.

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    val metric: Metric = createMetricBuilder(metricObj.name)
+
+    // Prepares the time series request
+    val timeSeries = createTimeSeries(metricObj.kind, metric, monitoredResource, dataPointList.asJava)
+    val timeSeriesList = List[TimeSeries](timeSeries)
+
+    val timeSeriesRequest = CreateTimeSeriesRequest.newBuilder.setName(projectName.toString).addAllTimeSeries(timeSeriesList.asJava).build
+
+    // Writes time series data
+    metricServiceClient.createTimeSeries(timeSeriesRequest)
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: TOL:

The javadoc for this method says:
```
   * @throws com.google.api.gax.rpc.ApiException if the remote call fails
```

If that happens then the actor will crash and - since this is probably using the default strategy - be restarted with its state reset to nothing (ie it would have no data in the metricMap).

That might be what you want (since it's the simplest answer and it's not the end of the world if it does happen) - but you might want to handle these exceptions more gracefully?

File: services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfigSpec.scala [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    stackdriverConfig.auth.name shouldBe "application-default"
+    stackdriverConfig.googleProject shouldBe "my-project"
+    stackdriverConfig.flushRate shouldBe 1.minute
+    stackdriverConfig.cromwellInstanceIdentifier shouldBe Some("cromwell-101")
+    stackdriverConfig.cromwellInstanceRole shouldBe Some("backend")
+    stackdriverConfig.cromwellPerfTestCase shouldBe Some("perf-test-1")
+  }
+
+
+  it should "coorectly parse config with optional values" in {
```
Comment (mcovarr) [inline services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfigSpec.scala]:
```suggestion
  it should "correctly parse config with optional values" in {
```

File: services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfigSpec.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        |auth = "my-auth"
+        |google-project = "my-project"
+        |flush-rate = 1 minute
+      """.stripMargin
+    )
+
+    val exception = the[IllegalArgumentException] thrownBy StackdriverConfig(config, googleConfig)
+
+    exception.getMessage shouldBe """Stackdriver instrumentation config is invalid.:
+                                    |`auth` scheme is invalid. Errors: NonEmptyList(`google` configuration stanza does not contain an auth named 'my-auth'.  Known auth names: application-default)""".stripMargin
```
Comment (mcovarr) [inline services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfigSpec.scala]: TOL that stringification of the errors NEL is less than ideal but probably a preexisting issue...

File: docs/developers/Instrumentation.md [hunk 7]
```diff
@@ -51,3 +52,47 @@ These are the current high level categories:
 * `job`
 * `workflow`
 * `io`
+
+
+###Stackdriver
+
+Cromwell now also supports sending metrics to [Google's Stackdriver API](https://cloud.google.com/monitoring/api/v3/). To instantiate the Stackdriver instrumentation service actor
```
[the remark this answers is not in the document]
Comment (cjllanwarne) [reply docs/developers/Instrumentation.md]: Interesting... I almost commented in the opposite direction. ie to say that I would prefer "Cromwell also supports sending metrics [...]"

File: cromwell.example.backends/cromwell.examples.conf [hunk 8]
```diff
@@ -501,6 +501,21 @@ services {
     #   prefix = "" # can be used to prefix all metrics with an api key for example
     #   flush-rate = 1 second # rate at which aggregated metrics will be sent to statsd
     # }
+
+    # Stackdriver- Send metrics to Google's monitoring API
+    # class = "cromwell.services.instrumentation.impl.stackdriver.StackdriverInstrumentationServiceActor"
+    # config {
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply cromwell.example.backends/cromwell.examples.conf]: Now that I think of it, since one can only instantiate either StatsD or Stackdriver for instrumentation, I will switch `config.statsd` to simply `config`. How does that sound?

File: cromwell.example.backends/cromwell.examples.conf [hunk 9]
```diff
@@ -501,6 +501,21 @@ services {
...
+    #   # auth scheme can be `application_default` or `service_account`
+    #   auth = "service-account"
+    #   google-project = "my-project"
+    #   # rate at which aggregated metrics will be sent to Stackdriver. It needs to be equal or greater than 1 minute.
+    #   # Google's Stackdriver API needs each metric to be sent not more than once per minute.
+    #   flush-rate = 1 minute
+    #   # below 3 keys are attached as labels to each metric. `cromwell-perf-test-case` is specifically meant for perf env.
+    #   cromwell-instance-identifier = "cromwell-101"
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply cromwell.example.backends/cromwell.examples.conf]: The top level config already has `system.cromwell_id`. I had thought about reusing it, but I don't remember why I didn't. I will give that a try.

File: docs/developers/Instrumentation.md [hunk 10]
```diff
@@ -1,14 +1,15 @@
-**Overview**
+#Overview
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply docs/developers/Instrumentation.md]: Ah I see! Will add it.

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    val cromwellInstanceIdentifier: ErrorOr[Option[String]] = serviceConfig.as[Option[String]](CromwellInstanceIdentifier).validNel
+    val cromwellInstanceRole: ErrorOr[Option[String]] = serviceConfig.as[Option[String]](CromwellInstanceRole).validNel
+    val cromwellPerfTestCase: ErrorOr[Option[String]] = serviceConfig.as[Option[String]](CromwellPerfTest).validNel
+
+    (googleProject, authScheme, flushRate, cromwellInstanceIdentifier, cromwellInstanceRole, cromwellPerfTestCase).mapN({ (p, a, f, i, r, t) =>
+      new StackdriverConfig(p, a, f, i, r, t)
+    }).valueOr(errors => throw new IllegalArgumentException with MessageAggregation {
+      override val exceptionContext = "Stackdriver instrumentation config is invalid."
+      override val errorMessages = errors.toList
+    })
```
Comment (kshakir) [inline services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala]: Slightly different (`Exception` vs. `IAE`) but more compact (after switching an `import`):
```scala
}).valueOr(errors =>
  throw AggregatedMessageException("Stackdriver instrumentation config is invalid.", errors.toList)
)
```

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+    labelsList.toMap
+  }
+
+
+  private def updateMetricMap(bucket: CromwellBucket, metricValue: Double, metricKind: StackdriverMetricKind): Unit = {
+    val metricObj = StackdriverMetric(bucket.toStackdriverString, metricKind)
+
+    if (metricsMap.contains(metricObj)) {
+      val valueList = metricsMap(metricObj)
+      metricsMap += metricObj ->  valueList.::(metricValue)
```
Comment (kshakir) [inline services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]:
```suggestion
      metricsMap += metricObj -> valueList :: metricValue
```

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+    }
+    else metricsMap += metricObj -> List(metricValue)
```
[the remark this answers is not in the document]
Comment (kshakir) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: List prepend would be faster, Vector is "generally simpler". See also: https://docs.scala-lang.org/overviews/collections/performance-characteristics.html

File: services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfigSpec.scala [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    )
+
+    val stackdriverConfig = StackdriverConfig(config, googleConfig)
+
+    stackdriverConfig.auth.name shouldBe "application-default"
+    stackdriverConfig.googleProject shouldBe "my-project"
+    stackdriverConfig.flushRate shouldBe 1.minute
+    stackdriverConfig.cromwellInstanceIdentifier shouldBe Some("cromwell-101")
+    stackdriverConfig.cromwellInstanceRole shouldBe Some("backend")
+    stackdriverConfig.cromwellPerfTestCase shouldBe Some("perf-test-1")
```
Comment (kshakir) [inline services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfigSpec.scala]: `Some.apply()` when used incorrectly causes very strange, very hard to debug problems. As `Option.apply()` can do the same exact thing as `Some.apply()` with a tiny performance penalty, the latter is considered banned from the codebase.

TL;DR Use `Option()` aka `Option.apply()`.

File: cromwell.example.backends/cromwell.examples.conf (hunk 9 above)
[the remark this answers is not in the document]
Comment (salonishah11) [reply cromwell.example.backends/cromwell.examples.conf]: I am reusing `system.cromwell_id` rather than creating a new key `cromwell-instance-identifier`.

File: project/Dependencies.scala [hunk 15]
```diff
@@ -601,6 +606,19 @@ object Dependencies {
...
+    "grpc-alts",
+    "grpc-auth",
+    "grpc-context",
+    "grpc-core",
+    "grpc-grpclb",
+    "grpc-netty-shaded",
+    "grpc-protobuf-lite",
+    "grpc-protobuf",
+    "grpc-stub",
+  ).map(m => "io.grpc" % m % grpcV)
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply project/Dependencies.scala]: Right! Thank you!

File: cromwell.example.backends/cromwell.examples.conf (hunk 8 above)
[the remark this answers is not in the document]
Comment (cjllanwarne) [reply cromwell.example.backends/cromwell.examples.conf]: Sounds good to me!

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala (hunk 1 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: Did you mean that I should return `Option[List[(String, String)]]`?

Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: Or just `Option[(String, String)]`?

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+      case Invalid(e) => e.invalid
+    }
+  }
+
+
+  private def validateAuth(authSchemeString: String, googleConfiguration: GoogleConfiguration): ErrorOr[GoogleAuthMode] = {
+    validate[String](authSchemeString) match {
+      case Valid(schemeString) => googleConfiguration.auth(schemeString) match {
+        case Valid(_: UserServiceAccountMode) => s"`auth` scheme: $schemeString is of type `user_service_account` which is not allowed for Stackdriver instrumentation.".invalidNel
+        case Valid(auth) => auth.valid
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala]: The other 2 auths I see are `user_account` and `refreshToken`. Based on what `user_account` expects, I think it should work. I am not sure what exactly is auth of type `refreshToken`. I will take a look.

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala (hunk 11 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala]: Is it more meaningful to throw just `AggregatedMessageException ` over `IllegalArgumentException & MessageAggregation `? I thought since the the config values are wrong, it would make more sense for it to have `IllegalArgumentException`, no?

File: services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfigSpec.scala (hunk 14 above)
Comment (salonishah11) [reply services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfigSpec.scala]: Will change it to `Option(...)`. Thanks!

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala (hunk 1 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: Talked F2F, it works with `Option[(String, String)]`

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala (hunk 2 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: TIL. Thank you!

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: List prepend and Vector both would be equally efficient here, but since Vector seems to more simple, I will go with that.

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+object StackdriverConfig {
+  val CromwellInstanceIdentifier = "cromwell-instance-identifier"
+  val CromwellInstanceRole = "cromwell-instance-role"
+  val CromwellPerfTest = "cromwell-perf-test-case"
+
+  private def validateFlushRate(rate: FiniteDuration): ErrorOr[FiniteDuration] = {
+    validate[FiniteDuration](rate) match {
+      case Valid(duration) => duration match {
+        case d: FiniteDuration if d.<(1.minute) => (s"`flush-rate` can't be less than 1 minute. Current rate is `${d.toString}`. " +
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala]: @mcovarr if the config is invalid, the error returned is `Stackdriver instrumentation config is invalid. Errors(s): `flush-rate`...`. It would be redundant to add `*Stackdriver instrumentation* `flush-rate`` again in the error message. I have updated the remaining part of the error message from your suggestion.

@kshakir I have updated the logic with your suggestions.

Thank you both!

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala (hunk 4 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: That's a good point. I will handle the exception.

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala]: Oh I completely missed the part where you mentioned `new auth schemes`. Can a user create new auth schemes?

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala (hunk 11 above)
Comment (kshakir) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfig.scala]: Technically yes, but our Exception handling is a mixed bag. AFAIK we don't (currently) catch nor pay attention to the type of the exceptions.

File: services/src/main/scala/cromwell/services/instrumentation/impl/statsd/StatsDConfig.scala [hunk 18]
```diff
@@ -15,12 +15,11 @@ case class StatsDConfig(hostname: String, port: Int, prefix: Option[String], flu
...
   def apply(serviceConfig: Config): StatsDConfig = {
-    val statsDConfig = serviceConfig.getConfig("statsd")
 
-    val hostname: ErrorOr[String] = validate[String] { statsDConfig.as[String]("hostname") }
-    val port: ErrorOr[Int] = validate[Int] { statsDConfig.as[Int]("port") }
-    val prefix: ErrorOr[Option[String]] = statsDConfig.as[Option[String]]("prefix").validNel
-    val flushRate: ErrorOr[FiniteDuration] = validate[FiniteDuration] { statsDConfig.as[FiniteDuration]("flush-rate") }
+    val hostname: ErrorOr[String] = validate[String] { serviceConfig.as[String]("hostname") }
+    val port: ErrorOr[Int] = validate[Int] { serviceConfig.as[Int]("port") }
+    val prefix: ErrorOr[Option[String]] = serviceConfig.as[Option[String]]("prefix").validNel
```
Comment (kshakir) [inline services/src/main/scala/cromwell/services/instrumentation/impl/statsd/StatsDConfig.scala]:
```suggestion
    val prefix: ErrorOr[Option[String]] = validate { serviceConfig.as[Option[String]]("prefix") }
```

Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/statsd/StatsDConfig.scala]: Oh I missed that last time! Thank you.

**Comment by salonishah11** [comment]
@cjllanwarne @kshakir @mcovarr this is ready for review round 2!

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+
+  private def sendMetricData(): Unit = {
+    metricsMap.foreach { case (key, value: Vector[Double]) =>
+      val dataPointVectorSum = value.sum
+      val dataPoint = key.kind match {
+        case StackdriverGauge => dataPointVectorSum / value.length
+        case StackdriverCumulative => dataPointVectorSum
+      }
+
+      writeMetrics(key, dataPoint) match {
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: Oh yes. TIL. I will update that to recover. Thank you!

File: services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+      metricKind match {
+        case StackdriverGauge => TimeInterval.newBuilder.setEndTime(Timestamps.fromMillis(System.currentTimeMillis)).build
+        case StackdriverCumulative => TimeInterval.newBuilder.setStartTime(ActorCreationTime).setEndTime(Timestamps.fromMillis(System.currentTimeMillis)).build
+      }
+    }
+
+    def createMetricBuilder(metricName: String): Metric = {
+      metricLabelsMap match {
+        case map: Map[String, String] if map.isEmpty => Metric.newBuilder.setType(s"$CustomMetricDomain/$metricName").build
+        case map: Map[String, String] if map.nonEmpty => Metric.newBuilder.setType(s"$CustomMetricDomain/$metricName").putAllLabels(map.asJava).build
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala]: Actually I checked, we can pass in empty labels map to each metric and it doesn't throw any errors. Although now I am not sure why I had assumed that in the first place! So I have removed the pattern matching for `metricLabelsMap` altogether.

File: CHANGELOG.md [hunk 21]
```diff
@@ -1,5 +1,18 @@
...
+### Stackdriver Instrumentation
+
+Cromwell now supports sending metrics to [Google's Stackdriver API](https://cloud.google.com/monitoring/api/v3/). 
+Learn more on how to configure [here](https://cromwell.readthedocs.io/en/stable/developers/Instrumentation/).
+
+### StatsD Instrumentation Config change
+
+There is a small change in StatsD's configuration path. Originally, the path to the config was `services.Instrumentation.config.statsd`
+which now has been updated to `services.Instrumentation.config`. More info on its configuration can be found
+[here](https://cromwell.readthedocs.io/en/stable/developers/Instrumentation/).
```
Comment (cjllanwarne) [inline CHANGELOG.md]: TOL: Not a big deal in this case but if we end up with a *lot* of these duplicated link strings, we might want to switch to "reference" style links (https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links) where each URL is written down as a reference and can be linked to multiple times from within the text

File: services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfigSpec.scala (hunk 6 above)
Comment (cjllanwarne) [reply services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverConfigSpec.scala]: IMO, as long as it's not too hard to trace, it'd be nice to make that error message friendlier.

File: services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActorSpec.scala [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+      CromwellCount(testBucket, 80),
+      CromwellGauge(testGaugeBucket, 20),
+      CromwellTiming(testBucket.expand("timing"), 5.seconds),
+      CromwellIncrement(testBucket),
+      CromwellCount(testBucket, 80),
+      CromwellGauge(testGaugeBucket, 60),
+      CromwellTiming(testBucket.expand("timing"), 10.seconds)
+    )
+
+    val receivedCumulativeMetrics = ("custom.googleapis.com/cromwell/test_prefix/test/metric/bucket", 162.0)
```
Comment (cjllanwarne) [inline services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActorSpec.scala]: is this `expectedReceivedCumulativeMetrics`?

File: services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/TestStackdriverInstrumentationServiceActor.scala [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+class TestStackdriverInstrumentationServiceActor(serviceConfig: Config, globalConfig: Config, serviceRegistryActor: ActorRef)
+  extends StackdriverInstrumentationServiceActor(serviceConfig, globalConfig, serviceRegistryActor) {
+
+  var metricsReceived = List[TimeSeriesRequest]()
+
+  override def sendTimeSeriesToStackdriver(timeSeriesRequest: CreateTimeSeriesRequest) = {
+    val timeSeries = timeSeriesRequest.getTimeSeries(0)
+    val metric = timeSeries.getMetric
+
+    val abc = TimeSeriesRequest(metric.getType,
```
Comment (cjllanwarne) [inline services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/TestStackdriverInstrumentationServiceActor.scala]: `abc`?

Comment (salonishah11) [reply services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/TestStackdriverInstrumentationServiceActor.scala]: Ugh. I forgot to clean this up. Will do it now.

Comment (salonishah11) [reply services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/TestStackdriverInstrumentationServiceActor.scala]: Ugh. I forgot to clean this up. Will do it now.

File: services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActorSpec.scala (hunk 22 above)
Comment (salonishah11) [reply services/src/test/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActorSpec.scala]: Yes. I can rename it to `expectedCumulativeMetrics` if it makes more sense?
